content/post/smaller-the-parameter-smaller-the-bugs.md

```diff
@@ -0,0 +1,243 @@
+
+```rb
+class Book
+  attr_reader :title, :publish_year, :author
+  
+  def initialize(title, publish_year, author)
+    @title = title
+    @publish_year = publish_year
+    @author = author
+  end
+ 
+  def summary
+    "#{title} - #{author.full_name} (#{publish_year})"
+  end
+  
+  def cover
+    "#{title}\n\n\n#{author.full_name}"
+  end
+end
+  
+bater = Author.new("Bater", "Chen")
+book = Book.new("Ruby is the best", "2020", bater)
+book.summary
+=> "Ruby is the best - Chen, Bater (2020)"
+book.cover
+=> "Ruby is the best\n\n\nChen, Bater"
+```
+Both `summary` and `cover` method's output remain the same, but now we can call it freely without any parameter once the book object is prepared. In ruby, we can skip parentheses if there is no parameter.
+
+This example doesn't encourage you to move all the parameters into the object attributes, it depends on the domain know-how. We can do it because it makes sense for a book object to having an author as an attribute in this case. It may get worse if we try to put the wrong data as the object attribute.
+
+## Open-closed principle
+Reducing the parameter from two to zero is not the only benefit we got from this refactoring, this new design is open for extension but closed for modification. It follows the [open-closed principle][3].
+
+Now we have a new requirement coming again, some authors have the middle name. In the new version, it would be very easy to adapt to the change, we don't need to touch the book class anymore. The new author class would refactor like this:
+
+```rb
+class Author
+  attr_reader :first_name, :last_name, :middle_name
+  
+  def initialize(first_name, last_name, middle_name = nil)
+    @first_name = first_name
+    @last_name = last_name
+    @middle_name = middle_name
+  end
+  
+  def full_name
+    if middle_name.nil?
+      "#{last_name}, #{first_name}"
+    else
+      "#{last_name}, #{middle_name}, #{first_name}"
+    end
+  end
+end
+  
+bater = Author.new("Bater", "Chen")
+Book.new("Ruby is the best", "2020", bater).summary
+=> "Ruby is the best - Chen, Bater (2020)"
+  
+robert = Author.new("Robert", "Kiyosaki", "T")
+Book.new("Rich Dad, Poor Dad", "1997", robert).summary
+=> "Rich Dad, Poor Dad - Kiyosaki, T, Robert (1997)"
+```
+In this design, the existing behavior doesn't change. The default value of the middle name is `nil` (Null) because in most cases we can skip it, and we won't break anything when the existing code tries to create a new author with only two parameters. The author class is open for extension and the book class is closed for modification in this case. Without this refactoring, we need to touch on both `summary` and `cover` methods, and the parameters increase from two to three. Let's compare the code before and after the refactoring again:
```

Comment:
```suggestion
In this design, the existing behavior doesn't change. The default value of the middle name is `nil` (also known as Null in other languages) because in most cases we can skip it and we won't break anything when the existing code tries to create a new author with only two parameters. The author class is open for extension and the book class is closed for modification in this case. Without refactoring, we need to touch upon both the `summary` and `cover` methods, causing the parameters to increase from two to three. Let's compare the code before and after, again:
```